src/pipeline/hashmap_thread.rs

```diff
@@ -0,0 +1,65 @@
+use crate::pipeline::error::Res;
+use prost::alloc::vec::Vec as ProstVec;
+use std::collections::HashMap;
+use tokio::sync::{mpsc, oneshot};
+use uuid::Uuid;
+
+#[derive(Debug)]
+pub enum HashMapCommand {
+    Write(Uuid, ProstVec<u8>, oneshot::Sender<Option<ProstVec<u8>>>),
+    Remove(Uuid, oneshot::Sender<Option<ProstVec<u8>>>),
+}
+pub struct HashMapHandler {
+    name: &'static str,
+    m: HashMap<Uuid, ProstVec<u8>>,
+    receiver: mpsc::Receiver<HashMapCommand>,
+}
+impl HashMapHandler {
+    #[must_use]
+    pub fn new(name: &'static str, receiver: mpsc::Receiver<HashMapCommand>) -> HashMapHandler {
+        HashMapHandler {
+            name,
+            m: HashMap::new(),
+            receiver,
+        }
+    }
+
+    pub async fn run(mut self) {
+        while let Some(command) = self.receiver.recv().await {
+            let res = match command {
+                HashMapCommand::Write(key, value, ack) => self.write(key, value, ack).await,
+                HashMapCommand::Remove(key, ack) => self.remove(key, ack).await,
+            };
+            if res.is_err() {
+                println!(
+                    "{} could not complete operation on HashMap: {}",
+                    self.name,
+                    res.unwrap_err()
+                );
+            }
+        }
+    }
+    async fn write(
+        &mut self,
+        key: Uuid,
+        value: ProstVec<u8>,
+        ack: oneshot::Sender<Option<ProstVec<u8>>>,
+    ) -> Res<()> {
+        println!("{} writing data with key {key}", self.name);
+        let ousted = self.m.insert(key, value);
```

Comment:
insert returns either old value or `None` if `key` is not in the map, so I would expect this code to send back `None` most of the time. Looking at this code, I am not sure I understand why we need ack channel at all. Is it some sort of backpressure mechanism or would it be sufficient for clients to await `send` method on their side to move on? That would simplify the communication protocol

Reply:
sure. i was returning it just in case caller wanted to know that info, but it seems unlikely they would